cpp/demo/poisson_matrix_free/main.cpp

```diff
@@ -0,0 +1,215 @@
+// Copyright (C) 2022 Igor A. Baratta
+//
+// This file is part of DOLFINx (https://www.fenicsproject.org)
+//
+// SPDX-License-Identifier:    LGPL-3.0-or-later
+
+// =====================================
+// Matrix-free Conjugate Gradient solver
+// =====================================
+//
+// This demo illustrates how to:
+// * Solve a linear partial differential equation using a matrix free CG solver
+// * Create and apply Dirichlet boundary conditions
+// * Compute errors
+//
+// .. math::
+//    - \nabla^{2} u &= f \quad {\rm in} \ \Omega, \\
+//      u &= u_D \quad {\rm on} \ \Gamma_{D}
+//
+//  Where
+// .. math::
+//    u_D &= 1 + x^2 + 2y^2, \\
+//    f = -6
+//
+
+#include "poisson.h"
+#include <cmath>
+#include <dolfinx.h>
+#include <dolfinx/fem/Constant.h>
+#include <xtensor/xarray.hpp>
+
+using namespace dolfinx;
+
+namespace linalg
+{
+/// Compute vector r = alpha*x + y
+/// @param[in, out]  r Result
+/// @param[in]  alpha
+/// @param[in]  x
+/// @param[in]  y
+template <typename U>
+void axpy(la::Vector<U>& r, U alpha, const la::Vector<U>& x,
+          const la::Vector<U>& y)
+{
+  std::transform(x.array().cbegin(),
+                 std::next(x.array().cbegin(), x.map()->size_local()),
+                 y.array().cbegin(), r.mutable_array().begin(),
+                 [alpha](auto x, auto y) { return alpha * x + y; });
+}
+
+/// Solve problem A.x = b using the Conjugate Gradient method
+/// @tparam U The scalar type
```

Comment:
Is it possible to use T everywhere?

Reply:
It's cleaner to separate the two. Ideally (and hopefully soon) the demo will use multiple types since there is not dependency on PETSc. Issue is generated code for multiple types.